### PR TITLE
Dummy participant [12948]

### PR DESCRIPTION
**Pull request by jparisu**
Merge after #40

### CODE REVIEW

File: include/ddsrouter/participant/implementations/auxiliar/DummyParticipant.hpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    virtual std::shared_ptr<IReader> create_reader(
+            RealTopic topic) override;
+
+    virtual void add_discovered_endpoint(const Endpoint& new_endpoint);
+
+    virtual Endpoint get_discovered_endpoint(const Guid& guid) const;
+
+    void add_message_to_send(RealTopic topic, DataToSend data);
+
+    std::vector<DataStoraged> data_received_ref(RealTopic topic);
```
Comment (juanlofer-eprosima) [inline include/ddsrouter/participant/implementations/auxiliar/DummyParticipant.hpp:55]:
```suggestion
    std::vector<DataStored> data_received_ref(RealTopic topic);
```

File: src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+void DummyParticipant::add_message_to_send(RealTopic topic, DataToSend data)
+{
+    auto it = readers_.find(topic);
+    if (it == readers_.end())
+    {
+        it->second->add_message_to_send(data);
+    }
+}
+
+std::vector<DataStoraged> DummyParticipant::data_received_ref(RealTopic topic)
```
Comment (juanlofer-eprosima) [inline src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp:81]:
```suggestion
std::vector<DataStored> DummyParticipant::data_received_ref(RealTopic topic)
```

File: src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+{
+    auto it = writers_.find(topic);
+    if (it == writers_.end())
+    {
+        return it->second->data_received_ref();
+    }
+    else
+    {
+        return std::vector<DataStoraged>();
```
Comment (juanlofer-eprosima) [inline src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp:90]:
```suggestion
        return std::vector<DataStored>();
```

File: src/cpp/writer/implementations/auxiliar/DummyWriter.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    : EchoWriter(participant_id, topic)
+{
+}
+
+ReturnCode DummyWriter::write(std::unique_ptr<DataReceived>& data)
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    // Fill the data to store
+    DataStoraged new_data_to_store;
```
Comment (juanlofer-eprosima) [inline src/cpp/writer/implementations/auxiliar/DummyWriter.cpp:36]:
```suggestion
    DataStored new_data_to_store;
```

File: src/cpp/writer/implementations/auxiliar/DummyWriter.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    new_data_to_store.timestamp = now();
+    new_data_to_store.guid_src = data->source_guid;
+    new_data_to_store.payload = data->data;
+
+    data_storaged.push_back(new_data_to_store);
+
+    return ReturnCode::RETCODE_OK;
+}
+
+std::vector<DataStoraged> DummyWriter::data_received_ref()
```
Comment (juanlofer-eprosima) [inline src/cpp/writer/implementations/auxiliar/DummyWriter.cpp:46]:
```suggestion
std::vector<DataStored> DummyWriter::data_received_ref()
```

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]
Please fix uncrustify test

File: include/ddsrouter/dynamic/DiscoveryDatabase.hpp [hunk 6]
```diff
@@ -52,6 +52,8 @@ class DiscoveryDatabase
     ReturnCode erase_endpoint(
             const Endpoint& endpoint_to_erase);
 
+    Endpoint get_endpoint(const Guid& guid_of_endpoint_to_erase) const;
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/dynamic/DiscoveryDatabase.hpp:55]: :+1:

File: include/ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+#include <ddsrouter/writer/IWriter.hpp>
+#include <ddsrouter/writer/implementations/auxiliar/EchoWriter.hpp>
+#include <ddsrouter/types/ParticipantId.hpp>
+#include <ddsrouter/types/Timestamp.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+
+struct DataStoraged
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp:32]: :+1:

File: include/ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+public:
+
+    DummyWriter(
+        const ParticipantId& participant_id,
+        const RealTopic& topic);
+
+    ReturnCode write(
+            std::unique_ptr<DataReceived>& data) override;
+
+    std::vector<DataStoraged> data_received_ref();
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp:53]: :+1:

File: include/ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+protected:
+
+    std::vector<DataStoraged> data_storaged;
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp:57]: :+1:

File: include/ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+#ifndef _DATABROKER_WRITER_IMPLEMENTATIONS_AUX_DUMMYWRITER_HPP_
+#define _DATABROKER_WRITER_IMPLEMENTATIONS_AUX_DUMMYWRITER_HPP_
+
+#include <mutex>
+
+#include <ddsrouter/writer/IWriter.hpp>
+#include <ddsrouter/writer/implementations/auxiliar/EchoWriter.hpp>
+#include <ddsrouter/types/ParticipantId.hpp>
+#include <ddsrouter/types/Timestamp.hpp>
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp:27]: :+1:

File: src/cpp/dynamic/DiscoveryDatabase.cpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+ReturnCode DiscoveryDatabase::erase_endpoint(
+        const Endpoint& endpoint_to_erase)
+{
+    // TODO
+    throw UnsupportedException("DiscoveryDatabase::erase_endpoint is not implemented yet");
+}
+
+Endpoint DiscoveryDatabase::get_endpoint(
+    const Guid& guid_of_endpoint_to_erase) const
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/dynamic/DiscoveryDatabase.cpp:62]: :+1:

File: src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+// limitations under the License.
+
+/**
+ * @file DummyParticipant.cpp
+ */
+
+#include <ddsrouter/participant/implementations/auxiliar/DummyParticipant.hpp>
+#include <ddsrouter/reader/implementations/auxiliar/DummyReader.hpp>
+#include <ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp>
+#include <ddsrouter/types/ParticipantType.hpp>
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp:22]: :+1:

File: src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+Endpoint DummyParticipant::get_discovered_endpoint(const Guid& guid) const
+{
+    return discovery_database_->get_endpoint(guid);
+}
+
+void DummyParticipant::add_message_to_send(RealTopic topic, DataToSend data)
+{
+    auto it = readers_.find(topic);
+    if (it == readers_.end())
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp:75]: :+1:

File: src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        it->second->add_message_to_send(data);
+    }
+}
+
+std::vector<DataStoraged> DummyParticipant::data_received_ref(RealTopic topic)
+{
+    auto it = writers_.find(topic);
+    if (it == writers_.end())
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp:84]: :+1:

File: src/cpp/reader/implementations/auxiliar/DummyReader.cpp [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    , topic_(topic)
+{
+}
+
+void DummyReader::enable()
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+    enabled_.store(true);
+
+    // If there are data pending to be sent, call on data available callback
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/reader/implementations/auxiliar/DummyReader.cpp:37]: :+1:

File: src/cpp/reader/implementations/auxiliar/DummyReader.cpp [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    // If it is not enabled, cannot retrieve data
+    if (!enabled_.load())
+    {
+        return ReturnCode::RETCODE_NOT_ENABLED;
+    }
+
+    // There are no data pending sent
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/reader/implementations/auxiliar/DummyReader.cpp:65]: :+1:

File: src/cpp/reader/implementations/auxiliar/DummyReader.cpp [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    data_received->data = next_data_to_send.payload;
+
+    return ReturnCode::RETCODE_OK;
+}
+
+void DummyReader::add_message_to_send(DataToSend data)
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    // Even in disabled, the data will be stored
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/reader/implementations/auxiliar/DummyReader.cpp:86]: :+1:

File: src/cpp/types/ParticipantType.cpp [hunk 18]
```diff
@@ -41,6 +41,11 @@ ParticipantType ParticipantTypeFactory::participant_type_from_name(
         // Echo type
         return ParticipantType::ECHO;
     }
+    else if (type == DUMMY_TYPE_NAME)
+    {
+        // Void type
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/types/ParticipantType.cpp:46]: :+1:

File: src/cpp/writer/implementations/auxiliar/DummyWriter.cpp [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    // Fill the data to store
+    DataStoraged new_data_to_store;
+    new_data_to_store.timestamp = now();
+    new_data_to_store.guid_src = data->source_guid;
+    new_data_to_store.payload = data->data;
+
+    data_storaged.push_back(new_data_to_store);
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/writer/implementations/auxiliar/DummyWriter.cpp:41]: :+1:

File: src/cpp/writer/implementations/auxiliar/DummyWriter.cpp [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+
+    return ReturnCode::RETCODE_OK;
+}
+
+std::vector<DataStoraged> DummyWriter::data_received_ref()
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+    return data_storaged;
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/writer/implementations/auxiliar/DummyWriter.cpp:49]: :+1:

File: include/ddsrouter/participant/implementations/auxiliar/DummyParticipant.hpp (hunk 1 above)
Comment (jparisu) [reply include/ddsrouter/participant/implementations/auxiliar/DummyParticipant.hpp:55]: :+1:

File: src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp (hunk 2 above)
Comment (jparisu) [reply src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp:81]: :+1:

File: src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp (hunk 3 above)
Comment (jparisu) [reply src/cpp/participant/implementations/auxiliar/DummyParticipant.cpp:90]: :+1:

File: src/cpp/writer/implementations/auxiliar/DummyWriter.cpp (hunk 4 above)
Comment (jparisu) [reply src/cpp/writer/implementations/auxiliar/DummyWriter.cpp:36]: :+1:

File: src/cpp/writer/implementations/auxiliar/DummyWriter.cpp (hunk 5 above)
Comment (jparisu) [reply src/cpp/writer/implementations/auxiliar/DummyWriter.cpp:46]: :+1:

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Comment by juanlofer-eprosima** [comment]
Some suggestions for the future:

- Consider a class hierarchy for data, relating DataReceived, DataToSend, DataStoraged and any other future data structs.
- Remove empty destructors and default constructors.
- Consider creating base classes implementing IParticipant, IReader and IWriter interfaces from which the rest of Participants/Readers/Writers would inherit from.

**Pull request merged**